File: admin-dashboard/src/utils/formatters.ts
// Utility functions for formatting data in the admin dashboard

// MARK: - Date Formatting
export function formatDate(dateString: string | null | undefined): string {
  if (!dateString) return 'Unknown'
  
  try {
    const date = new Date(dateString)
    if (isNaN(date.getTime())) return 'Invalid Date'
    
    return date.toLocaleDateString('en-US', {
      year: 'numeric',
      month: 'short',
      day: 'numeric'
    })
  } catch {
    return 'Invalid Date'
  }
}

export function formatDateTime(dateString: string | null | undefined): string {
  if (!dateString) return 'Unknown'
  
  try {
    const date = new Date(dateString)
    if (isNaN(date.getTime())) return 'Invalid Date'
    
    return date.toLocaleString('en-US', {
      year: 'numeric',
      month: 'short',
      day: 'numeric',
      hour: '2-digit',
      minute: '2-digit'
    })
  } catch {
    return 'Invalid Date'
  }
}

export function formatRelativeDate(dateString: string | null | undefined): string {
  if (!dateString) return 'Unknown'
  
  try {
    const date = new Date(dateString)
    if (isNaN(date.getTime())) return 'Invalid Date'
    
    const now = new Date()
    const diffInMs = now.getTime() - date.getTime()
    const diffInDays = Math.floor(diffInMs / (1000 * 60 * 60 * 24))
    
    if (diffInDays === 0) return 'Today'
    if (diffInDays === 1) return 'Yesterday'
    if (diffInDays < 7) return `${diffInDays} days ago`
    if (diffInDays < 30) return `${Math.floor(diffInDays / 7)} weeks ago`
    if (diffInDays < 365) return `${Math.floor(diffInDays / 30)} months ago`
    
    return `${Math.floor(diffInDays / 365)} years ago`
  } catch {
    return 'Invalid Date'
  }
}

// MARK: - File Size Formatting
export function formatFileSize(bytes: number | null | undefined): string {
  if (!bytes || bytes === 0) return 'Unknown'
  
  const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB']
  const i = Math.floor(Math.log(bytes) / Math.log(1024))
  
  if (i === 0) return `${bytes} ${sizes[i]}`
  
  return `${(bytes / Math.pow(1024, i)).toFixed(1)} ${sizes[i]}`
}

export function formatFileSizeDetailed(bytes: number | null | undefined): string {
  if (!bytes || bytes === 0) return 'Unknown'
  
  const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB']
  const i = Math.floor(Math.log(bytes) / Math.log(1024))
  
  if (i === 0) return `${bytes.toLocaleString()} ${sizes[i]}`
  
  return `${(bytes / Math.pow(1024, i)).toFixed(2)} ${sizes[i]} (${bytes.toLocaleString()} bytes)`
}

// MARK: - Price Formatting
export function formatPrice(price: string | number | null | undefined, currency: string = 'USD'): string {
  if (!price || price === '0' || price === 0) return 'Free'
  
  const priceNumber = typeof price === 'string' ? parseFloat(price) : price
  
  if (isNaN(priceNumber)) return 'Invalid Price'
  
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: currency
  }).format(priceNumber)
}

export function formatPriceRange(minPrice: string | number | null, maxPrice: string | number | null, currency: string = 'USD'): string {
  const minFormatted = formatPrice(minPrice, currency)
  const maxFormatted = formatPrice(maxPrice, currency)
  
  if (minFormatted === maxFormatted) return minFormatted
  
  return `${minFormatted} - ${maxFormatted}`
}

// MARK: - Rating Formatting
export function formatRating(rating: number | null | undefined): string {
  if (!rating || rating === 0) return 'No rating'
  return rating.toFixed(1)
}

export function formatRatingCount(count: number | null | undefined): string {
  if (!count || count === 0) return ''
  
  if (count >= 1_000_000) {
    return `${(count / 1_000_000).toFixed(1)}M`
  }
  
  if (count >= 1_000) {
    return `${(count / 1_000).toFixed(1)}K`
  }
  
  return count.toLocaleString()
}

export function formatRatingWithCount(rating: number | null | undefined, count: number | null | undefined): string {
  const ratingText = formatRating(rating)
  const countText = formatRatingCount(count)
  
  if (!countText) return ratingText
  
  return `${ratingText} (${countText})`
}

// MARK: - Text Formatting
export function truncateText(text: string, maxLength: number, suffix: string = '...'): string {
  if (!text || text.length <= maxLength) return text
  return text.substring(0, maxLength) + suffix
}

export function capitalizeFirstLetter(text: string): string {
  if (!text) return text
  return text.charAt(0).toUpperCase() + text.slice(1).toLowerCase()
}

export function formatAppName(name: string): string {
  return capitalizeFirstLetter(name)
}

export function formatDeveloperName(name: string): string {
  return capitalizeFirstLetter(name)
}

// MARK: - URL Formatting
export function isValidURL(urlString: string | null | undefined): boolean {
  if (!urlString) return false
  
  try {
    new URL(urlString)
    return true
  } catch {
    return false
  }
}

export function formatURL(urlString: string | null | undefined): string {
  if (!urlString || !isValidURL(urlString)) return ''
  
  try {
    const url = new URL(urlString)
    return url.hostname + url.pathname
  } catch {
    return urlString
  }
}

export function getDomainFromURL(urlString: string | null | undefined): string {
  if (!urlString || !isValidURL(urlString)) return ''
  
  try {
    const url = new URL(urlString)
    return url.hostname
  } catch {
    return ''
  }
}

// MARK: - Version Formatting
export function formatVersion(version: string | null | undefined): string {
  if (!version) return 'Unknown'
  return `v${version}`
}

export function compareVersions(version1: string, version2: string): number {
  const v1Parts = version1.split('.').map(Number)
  const v2Parts = version2.split('.').map(Number)
  
  const maxLength = Math.max(v1Parts.length, v2Parts.length)
  
  for (let i = 0; i < maxLength; i++) {
    const v1Part = v1Parts[i] || 0
    const v2Part = v2Parts[i] || 0
    
    if (v1Part < v2Part) return -1
    if (v1Part > v2Part) return 1
  }
  
  return 0
}

// MARK: - Status Formatting
export function formatStatus(status: string | null | undefined): string {
  if (!status) return 'Unknown'
  
  const statusMap: Record<string, string> = {
    'ACTIVE': 'Active',
    'PENDING': 'Pending',
    'INACTIVE': 'Inactive',
    'DRAFT': 'Draft'
  }
  
  return statusMap[status.toUpperCase()] || capitalizeFirstLetter(status)
}

export function getStatusColor(status: string | null | undefined): 'success' | 'warning' | 'error' | 'default' {
  if (!status) return 'default'
  
  const statusColors: Record<string, 'success' | 'warning' | 'error' | 'default'> = {
    'ACTIVE': 'success',
    'PENDING': 'warning',
    'INACTIVE': 'error',
    'DRAFT': 'default'
  }
  
  return statusColors[status.toUpperCase()] || 'default'
}

// MARK: - Source Formatting
export function formatSource(source: string | null | undefined): string {
  if (!source) return 'Unknown'
  
  const sourceMap: Record<string, string> = {
    'MAS': 'Mac App Store',
    'MAC_APP_STORE': 'Mac App Store',
    'CUSTOM': 'Custom',
    'IMPORTED': 'Imported'
  }
  
  return sourceMap[source.toUpperCase()] || capitalizeFirstLetter(source)
}

export function getSourceColor(source: string | null | undefined): 'primary' | 'secondary' | 'info' | 'default' {
  if (!source) return 'default'
  
  const sourceColors: Record<string, 'primary' | 'secondary' | 'info' | 'default'> = {
    'MAS': 'primary',
    'MAC_APP_STORE': 'primary',
    'CUSTOM': 'secondary',
    'IMPORTED': 'info'
  }
  
  return sourceColors[source.toUpperCase()] || 'default'
}

// MARK: - Number Formatting
export function formatNumber(num: number | null | undefined): string {
  if (num === null || num === undefined) return '0'
  return num.toLocaleString()
}

export function formatPercentage(value: number, total: number): string {
  if (total === 0) return '0%'
  return `${((value / total) * 100).toFixed(1)}%`
}

// MARK: - Time Formatting
export function formatDuration(seconds: number): string {
  if (seconds < 60) return `${seconds}s`
  if (seconds < 3600) return `${Math.floor(seconds / 60)}m ${seconds % 60}s`
  
  const hours = Math.floor(seconds / 3600)
  const minutes = Math.floor((seconds % 3600) / 60)
  return `${hours}h ${minutes}m`
}

// MARK: - Currency Formatting
export function formatCurrency(amount: number, currency: string = 'USD'): string {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: currency
  }).format(amount)
}

// MARK: - Validation Helpers
export function isValidEmail(email: string): boolean {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
  return emailRegex.test(email)
}

export function isValidPhoneNumber(phone: string): boolean {
  const phoneRegex = /^[\+]?[1-9][\d]{0,15}$/
  return phoneRegex.test(phone.replace(/[\s\-\(\)]/g, ''))
}

export function isValidSlug(slug: string): boolean {
  const slugRegex = /^[a-z0-9-]+$/
  return slugRegex.test(slug)
} 

File: ios/AppStoreDiscovery/AppStoreDiscovery/Utils/Formatters.swift
//
//  Formatters.swift
//  AppStoreDiscovery
//
//  Created by iMac on 13/7/2025.
//

import Foundation

// MARK: - String Extension for Regex
extension String {
    func matches(pattern: String) -> Bool {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else {
            return false
        }
        let range = NSRange(location: 0, length: self.count)
        return regex.firstMatch(in: self, options: [], range: range) != nil
    }
}

// MARK: - Date Formatter
struct DateFormatter {
    static let shared = DateFormatter()
    
    private let dateFormatter = Foundation.DateFormatter()
    
    func formatDate(_ dateString: String?, style: Foundation.DateFormatter.Style = .medium) -> String {
        guard let dateString = dateString else { return "Unknown" }
        
        // Try ISO format first
        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
        if let date = dateFormatter.date(from: dateString) {
            dateFormatter.dateStyle = style
            dateFormatter.timeStyle = .none
            return dateFormatter.string(from: date)
        }
        
        // Try simple date format
        dateFormatter.dateFormat = "yyyy-MM-dd"
        if let date = dateFormatter.date(from: dateString) {
            dateFormatter.dateStyle = style
            dateFormatter.timeStyle = .none
            return dateFormatter.string(from: date)
        }
        
        return dateString
    }
    
    func formatRelativeDate(_ dateString: String?) -> String {
        guard let dateString = dateString else { return "Unknown" }
        
        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
        guard let date = dateFormatter.date(from: dateString) else {
            return formatDate(dateString)
        }
        
        let now = Date()
        
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .abbreviated
        return formatter.localizedString(for: date, relativeTo: now)
    }
}

// MARK: - File Size Formatter
struct FileSizeFormatter {
    static func formatFileSize(_ size: String?) -> String {
        guard let size = size else { return "Unknown" }
        
        // If size is already in human-readable format (e.g., "97.4 MB"), return it as is
        if size.contains("MB") || size.contains("GB") || size.contains("KB") {
            return size
        }
        
        // Fallback: try to parse as integer and format
        if let sizeInt = Int(size) {
            let formatter = ByteCountFormatter()
            formatter.allowedUnits = [.useMB, .useGB]
            formatter.countStyle = .file
            return formatter.string(fromByteCount: Int64(sizeInt))
        }
        
        return size
    }
    
    static func formatFileSizeDetailed(_ size: String?) -> String {
        guard let size = size else { return "Unknown" }
        
        // If size is already in human-readable format, return it as is
        if size.contains("MB") || size.contains("GB") || size.contains("KB") {
            return size
        }
        
        // Fallback: try to parse as integer and format
        if let sizeInt = Int(size) {
            let formatter = ByteCountFormatter()
            formatter.allowedUnits = [.useKB, .useMB, .useGB]
            formatter.countStyle = .file
            return formatter.string(fromByteCount: Int64(sizeInt))
        }
        
        return size
    }
}

// MARK: - Price Formatter
struct PriceFormatter {
    static func formatPrice(_ price: String?, currency: String = "USD") -> String {
        guard let price = price, !price.isEmpty else { return "Free" }
        
        if price == "0" || price == "0.00" {
            return "Free"
        }
        
        // Try to format as currency
        if let priceValue = Double(price) {
            let formatter = NumberFormatter()
            formatter.numberStyle = .currency
            formatter.currencyCode = currency
            return formatter.string(from: NSNumber(value: priceValue)) ?? "$\(price)"
        }
        
        return "$\(price)"
    }
    
    static func formatPriceRange(_ minPrice: String?, maxPrice: String?, currency: String = "USD") -> String {
        let minFormatted = formatPrice(minPrice, currency: currency)
        let maxFormatted = formatPrice(maxPrice, currency: currency)
        
        if minFormatted == maxFormatted {
            return minFormatted
        }
        
        return "\(minFormatted) - \(maxFormatted)"
    }
}

// MARK: - Rating Formatter
struct RatingFormatter {
    static func formatRating(_ rating: Double?) -> String {
        guard let rating = rating else { return "No rating" }
        return String(format: "%.1f", rating)
    }
    
    static func formatRatingCount(_ count: Int?) -> String {
        guard let count = count else { return "" }
        
        if count >= 1_000_000 {
            return String(format: "%.1fM", Double(count) / 1_000_000)
        } else if count >= 1_000 {
            return String(format: "%.1fK", Double(count) / 1_000)
        } else {
            return "\(count)"
        }
    }
    
    static func formatRatingWithCount(_ rating: Double?, count: Int?) -> String {
        let ratingText = formatRating(rating)
        let countText = formatRatingCount(count)
        
        if countText.isEmpty {
            return ratingText
        }
        
        return "\(ratingText) (\(countText))"
    }
}

// MARK: - Text Formatter
struct TextFormatter {
    static func truncateText(_ text: String, maxLength: Int, suffix: String = "...") -> String {
        guard text.count > maxLength else { return text }
        let truncated = String(text.prefix(maxLength))
        return truncated + suffix
    }
    
    static func capitalizeFirstLetter(_ text: String) -> String {
        guard !text.isEmpty else { return text }
        return text.prefix(1).uppercased() + text.dropFirst()
    }
    
    static func formatAppName(_ name: String) -> String {
        return capitalizeFirstLetter(name)
    }
    
    static func formatDeveloperName(_ name: String) -> String {
        return capitalizeFirstLetter(name)
    }
}

// MARK: - URL Formatter
struct URLFormatter {
    static func isValidURL(_ urlString: String?) -> Bool {
        guard let urlString = urlString, !urlString.isEmpty else { return false }
        return URL(string: urlString) != nil
    }
    
    static func formatURL(_ urlString: String?) -> String {
        guard let urlString = urlString, isValidURL(urlString) else { return "" }
        
        // Remove protocol if present
        if urlString.hasPrefix("http://") {
            return String(urlString.dropFirst(7))
        } else if urlString.hasPrefix("https://") {
            return String(urlString.dropFirst(8))
        }
        
        return urlString
    }
    
    static func getDomainFromURL(_ urlString: String?) -> String {
        guard let urlString = urlString,
              let url = URL(string: urlString),
              let host = url.host else { return "" }
        
        return host
    }
}

// MARK: - Version Formatter
struct VersionFormatter {
    static func formatVersion(_ version: String?) -> String {
        guard let version = version, !version.isEmpty else { return "Unknown" }
        return "v\(version)"
    }
    
    static func compareVersions(_ version1: String, _ version2: String) -> ComparisonResult {
        let components1 = version1.components(separatedBy: ".").compactMap { Int($0) }
        let components2 = version2.components(separatedBy: ".").compactMap { Int($0) }
        
        let maxLength = max(components1.count, components2.count)
        
        for i in 0..<maxLength {
            let value1 = i < components1.count ? components1[i] : 0
            let value2 = i < components2.count ? components2[i] : 0
            
            if value1 < value2 {
                return .orderedAscending
            } else if value1 > value2 {
                return .orderedDescending
            }
        }
        
        return .orderedSame
    }
}

// MARK: - Category Formatter
struct CategoryFormatter {
    static func formatCategoryName(_ name: String) -> String {
        return TextFormatter.capitalizeFirstLetter(name)
    }
    
    static func getCategoryIcon(_ categoryName: String) -> String {
        let lowercased = categoryName.lowercased()
        
        switch lowercased {
        case let name where name.contains("productivity"):
            return "briefcase.fill"
        case let name where name.contains("game"):
            return "gamecontroller.fill"
        case let name where name.contains("photo"), let name where name.contains("video"):
            return "camera.fill"
        case let name where name.contains("music"), let name where name.contains("audio"):
            return "music.note"
        case let name where name.contains("social"):
            return "person.2.fill"
        case let name where name.contains("education"):
            return "book.fill"
        case let name where name.contains("health"), let name where name.contains("fitness"):
            return "heart.fill"
        case let name where name.contains("finance"), let name where name.contains("business"):
            return "dollarsign.circle.fill"
        case let name where name.contains("travel"):
            return "airplane"
        case let name where name.contains("food"), let name where name.contains("recipe"):
            return "fork.knife"
        case let name where name.contains("weather"):
            return "cloud.fill"
        case let name where name.contains("news"):
            return "newspaper.fill"
        default:
            return "app.fill"
        }
    }
}

// MARK: - Status Formatter
struct StatusFormatter {
    static func formatStatus(_ status: String?) -> String {
        guard let status = status else { return "Unknown" }
        
        switch status.uppercased() {
        case "ACTIVE":
            return "Active"
        case "PENDING":
            return "Pending"
        case "INACTIVE":
            return "Inactive"
        case "DRAFT":
            return "Draft"
        default:
            return TextFormatter.capitalizeFirstLetter(status)
        }
    }
    
    static func getStatusColor(_ status: String?) -> String {
        guard let status = status else { return "gray" }
        
        switch status.uppercased() {
        case "ACTIVE":
            return "green"
        case "PENDING":
            return "orange"
        case "INACTIVE":
            return "red"
        case "DRAFT":
            return "gray"
        default:
            return "gray"
        }
    }
}

// MARK: - Minimum OS Formatter
struct MinimumOSFormatter {
    static func formatMinimumOS(_ osVersion: String?) -> String {
        guard let osVersion = osVersion, !osVersion.isEmpty else { return "Unknown" }
        
        // Common patterns for macOS versions
        let lowercased = osVersion.lowercased()
        
        // Handle macOS version numbers
        if lowercased.contains("macos") || lowercased.contains("os x") {
            // Extract version number and format nicely
            let versionPattern = try? NSRegularExpression(pattern: "(\\d+\\.\\d+)", options: [])
            if let match = versionPattern?.firstMatch(in: osVersion, options: [], range: NSRange(location: 0, length: osVersion.count)) {
                let versionRange = Range(match.range(at: 1), in: osVersion)!
                let version = String(osVersion[versionRange])
                return "macOS \(version)+"
            }
            return osVersion
        }
        
        // Handle iOS version numbers
        if lowercased.contains("ios") {
            let versionPattern = try? NSRegularExpression(pattern: "(\\d+\\.\\d+)", options: [])
            if let match = versionPattern?.firstMatch(in: osVersion, options: [], range: NSRange(location: 0, length: osVersion.count)) {
                let versionRange = Range(match.range(at: 1), in: osVersion)!
                let version = String(osVersion[versionRange])
                return "iOS \(version)+"
            }
            return osVersion
        }
        
        // Handle simple version numbers (assume macOS)
        if osVersion.matches(pattern: "^\\d+\\.\\d+$") {
            return "macOS \(osVersion)+"
        }
        
        // Return as-is if no pattern matches
        return osVersion
    }
    
    static func getOSIcon(_ osVersion: String?) -> String {
        guard let osVersion = osVersion else { return "questionmark.circle" }
        
        let lowercased = osVersion.lowercased()
        
        if lowercased.contains("macos") || lowercased.contains("os x") {
            return "desktopcomputer"
        } else if lowercased.contains("ios") {
            return "iphone"
        } else {
            return "laptopcomputer"
        }
    }
}

// MARK: - Source Formatter
struct SourceFormatter {
    static func formatSource(_ source: String?) -> String {
        guard let source = source else { return "Unknown" }
        
        switch source.uppercased() {
        case "MAS", "MAC_APP_STORE":
            return "Mac App Store"
        case "CUSTOM":
            return "Custom"
        case "IMPORTED":
            return "Imported"
        default:
            return TextFormatter.capitalizeFirstLetter(source)
        }
    }
    
    static func getSourceIcon(_ source: String?) -> String {
        guard let source = source else { return "questionmark.circle" }
        
        switch source.uppercased() {
        case "MAS", "MAC_APP_STORE":
            return "app.store"
        case "CUSTOM":
            return "plus.circle"
        case "IMPORTED":
            return "arrow.down.circle"
        default:
            return "questionmark.circle"
        }
    }
} 